impl/src/main/java/org/ehcache/impl/internal/store/heap/OnHeapStore.java

```diff
@@ -647,29 +647,47 @@ public void clear() {
 
   @Override
   public Iterator<Cache.Entry<K, ValueHolder<V>>> iterator() {
+    java.util.Iterator<Entry<K, OnHeapValueHolder<V>>> iterator = map.entrySetIterator();
     return new Iterator<Cache.Entry<K, ValueHolder<V>>>() {
-      private final java.util.Iterator<Map.Entry<K, OnHeapValueHolder<V>>> it = map.entrySetIterator();
+      private Cache.Entry<K, ValueHolder<V>> prefetched = advance();
 
       @Override
       public boolean hasNext() {
-        return it.hasNext();
+        return prefetched != null;
       }
 
       @Override
-      public Cache.Entry<K, ValueHolder<V>> next() {
-        Entry<K, OnHeapValueHolder<V>> next = it.next();
-        K key = next.getKey();
-        OnHeapValueHolder<V> value = next.getValue();
-        return new Cache.Entry<K, ValueHolder<V>>() {
-          @Override
-          public K getKey() {
-            return key;
-          }
-          @Override
-          public ValueHolder<V> getValue() {
-            return value;
+      public Cache.Entry<K, ValueHolder<V>> next() throws StoreAccessException {
```

Comment:
Same prefetch comment as on `TieredStore`.

Reply:
You mean that it might be expired? Tough cheddar.